src/Bicep.Core/Text/SpellChecker.cs

```diff
@@ -0,0 +1,153 @@
+                }
+
+                if (checkCaseInsensitiveEqualityOnly)
+                {
+                    continue;
+                }
+
+                if (candidate.Length < 3)
+                {
+                    // Don't bother for such a short candidate name.
+                    continue;
+                }
+
+                int distance = GetLevenshteinDistanceWithMax(nameUpperCase, candidateUpperCase, maxDistance - 1);
+
+                if (distance < 0)
+                {
+                    continue;
+                }
+
+                if (distance < 2)
+                {
+                    /*
+                     * There won't be any better result with a smaller levenshiten distance,
+                     * so check case-insensitive equality for the rest of the candidates.
+                     */
+                    checkCaseInsensitiveEqualityOnly = true;
+                    bestCandidate = candidate;
+                }
+                else
+                {
+                    Debug.Assert(distance < maxDistance);
+
+                    maxDistance = distance;
+                    bestCandidate = candidate;
+                }
+            }
+
+            return bestCandidate;
+        }
+
+        private static int GetLevenshteinDistanceWithMax(string first, string second, int maxDistance)
+        {
+            if (Math.Abs(first.Length - second.Length) > maxDistance)
+            {
+                return -1;
+            }
+
+            /*
+             * Lev(i, j): levenshtein distance between first.Substring(0, i) and second.Substring(0, j).
+             * Using two 1D arrays instead of one 2d array to reduce space complexity.
+             */
+            var previous = new int[second.Length + 1];
+            var current = new int[second.Length + 1];
+
+            // Initialize Lev(0, j).
+            for (int j = 0; j <= second.Length; j++)
+            {
+                previous[j] = j;
+            }
+
+            // Start DP to fill Lev(i, j).
+            for (int i = 1; i <= first.Length; i++)
+            {
+                int minJ = i > maxDistance ? i - maxDistance : 1;
+                int maxJ = second.Length > maxDistance + i ? maxDistance + i : second.Length;
+                int minDistance = i;
+
+                // Lev(i, 0).
+                current[0] = i;
+
+                // Lev(i, 1) to Lev(i, minJ - 1).
+                for (int j = 1; j < minJ; j++)
+                {
+                    current[j] = int.MaxValue;
+                }
+
+                // Lev(i, minJ) to Lev(i, maxJ).
+                for (int j = minJ; j <= maxJ; j++)
+                {
+                    // See: https://en.wikipedia.org/wiki/Levenshtein_distance#Definition
+                    current[j] = first[i - 1] == second[j - 1]
+                        ? previous[j - 1]
+                        : 1 + Min(previous[j], current[j - 1], previous[j - 1]);

```

Comment:
I took the definition from Wikipedia, but TypeScript is using a slightly different version where it considers one substitution operation emits 2 edit distances (`previous[j - 1] + 1` instead of `previous[j - 1]`), maybe because that is equivalent to doing a deletion and a insertion.